### ISSUE TITLE
In case of consecutive vrrp state notification, state notification is not coming in correct order.

### ISSUE THREAD

**Issue by Navjot01** [acassen/keepalived#569]
For detail see below pull request and complete thread.

https://github.com/acassen/keepalived/pull/568

**Comment by Navjot01** [comment]
Backup ------------------------ Master
prio 100                                    prio 105

Now on backup side, i keep changing priority to 108 then back to 100. After changing priority number of times on backup side eventually i see both side calling the notify with state as "Master" i.s. last state update is Master even though KA maintains the correct state but notify is not called in correct order causing the problem.

**Comment by pqarmitage** [comment]
This isn't a solution to the underlying problem, but it might be a better solution to what you are doing, since reloading the configuration has quite an overhead. Whether it works for you depends on how rapidly you need to change the priority.

You could use a weighted track script with a weight of 8 which reads the exit code from a file, and then your interface script could, rather than restarting keepalived, write 0 or 1 to the file. So the track script could be something like
```
#!/bin/bash

ret=$(cat /etc/keepalived/ret/VI_0_prio.ret)
exit $ret
```
and in your keepalived config assuming the above script is `/etc/keepalived/scripts/VI_0_prio.sh` you would add:
```
vrrp_script prio_VI_0 {
    script /etc/keepalived/scripts/VI_0_prio.sh
    interval 1
    timeout 2
    rise 1
    fall 1
}
```
and in the config for the vrrp instance you would add
```
    track_script {
        prio_VI_0 weight 8
    }
```

If the VI_0_prio.ret file contents are `0` then the configured priority of the VI_0 vrrp instance will be increased by 8 (to 108), if the contents are `1` then the priority of the instance will be the configured priority (100). So then all your interface script needs to do is write 0 to 1 to the VI_0_prio.ret file.

Since the shortest interval between running scripts is 1 second, and then there is the delay before the next advert is sent, the new priority can't be guaranteed to be sent until at least script interval + advert interval seconds after the value is changed in the VI_0_prio.ret file.

**Comment by pqarmitage** [comment]
I have implemented a new approach with commit 04905cd et cf, and added a new mechanism for notifying events, which is to deliver them via a FIFO. keepalived will write a line to the FIFO for each event, which can then be read by a separate process. Since there is no asynchronous activity, the events will be delivered in the correct order, and this approach also avoids the overhead of forking a child for each notify event. An example script for reading the FIFO is provided in doc/samples/sample/sample_notify_fifo.sh.

The lines written to the FIFO are simply the parameters that are passed to a notify script, for example:
INSTANCE "VI_1" MASTER 100

The initial implementation has now been enhanced to include logging of LVS notify events, which can either be to the same or a different FIFO.

**Comment by pqarmitage** [comment]
If you want sequenced delivery of notifications, it is necessary to use notify_fifo rather than notify scripts.
